internal/datastore/common/changes.go

```diff
@@ -4,77 +4,181 @@ import (
 	return Changes[R, K]{
-		make(map[K]changeRecord[R], 0),
-		keyFunc,
+		records: make(map[K]changeRecord[R], 0),
+		keyFunc: keyFunc,
+		content: content,
 	}
 }
 
-// AddChange adds a specific change to the complete list of tracked changes
-func (ch Changes[R, K]) AddChange(
+// IsEmpty returns if the change set is empty.
+func (ch Changes[R, K]) IsEmpty() bool {
+	return len(ch.records) == 0
+}
+
+// AddRelationshipChange adds a specific change to the complete list of tracked changes
+func (ch Changes[R, K]) AddRelationshipChange(
 	ctx context.Context,
 	rev R,
 	tpl *core.RelationTuple,
 	op core.RelationTupleUpdate_Operation,
 ) {
+	if ch.content&datastore.WatchRelationships != datastore.WatchRelationships {
+		return
+	}
+
+	record := ch.recordForRevision(rev)
+	tplKey := tuple.StringWithoutCaveat(tpl)
+
+	switch op {
+	case core.RelationTupleUpdate_TOUCH:
+		// If there was a delete for the same tuple at the same revision, drop it
+		delete(record.tupleDeletes, tplKey)
+		record.tupleTouches[tplKey] = tpl
+
+	case core.RelationTupleUpdate_DELETE:
+		_, alreadyTouched := record.tupleTouches[tplKey]
+		if !alreadyTouched {
+			record.tupleDeletes[tplKey] = tpl
+		}
+	default:
+		log.Ctx(ctx).Fatal().Stringer("operation", op).Msg("unknown change operation")
+	}
+}
+
+func (ch Changes[R, K]) recordForRevision(rev R) changeRecord[R] {
 	k := ch.keyFunc(rev)
 	revisionChanges, ok := ch.records[k]
 	if !ok {
 		revisionChanges = changeRecord[R]{
 			rev,
 			make(map[string]*core.RelationTuple),
 			make(map[string]*core.RelationTuple),
+			make(map[string]datastore.SchemaDefinition),
+			make(map[string]struct{}),
+			make(map[string]struct{}),
 		}
 		ch.records[k] = revisionChanges
 	}
 
-	tplKey := tuple.StringWithoutCaveat(tpl)
+	return revisionChanges
+}
 
-	switch op {
-	case core.RelationTupleUpdate_TOUCH:
-		// If there was a delete for the same tuple at the same revision, drop it
-		delete(revisionChanges.tupleDeletes, tplKey)
+// AddDeletedNamespace adds a change indicating that the namespace with the name was deleted.
+func (ch Changes[R, K]) AddDeletedNamespace(
+	_ context.Context,
+	rev R,
+	namespaceName string,
+) {
+	if ch.content&datastore.WatchSchema != datastore.WatchSchema {
+		return
+	}
 
-		revisionChanges.tupleTouches[tplKey] = tpl
+	record := ch.recordForRevision(rev)
+	delete(record.definitionsChanged, "ns-"+namespaceName)
 
-	case core.RelationTupleUpdate_DELETE:
-		_, alreadyTouched := revisionChanges.tupleTouches[tplKey]
-		if !alreadyTouched {
-			revisionChanges.tupleDeletes[tplKey] = tpl
-		}
+	record.namespacesDeleted[namespaceName] = struct{}{}
+}
+
+// AddDeletedCaveat adds a change indicating that the caveat with the name was deleted.
+func (ch Changes[R, K]) AddDeletedCaveat(
+	_ context.Context,
+	rev R,
+	caveatName string,
+) {
+	if ch.content&datastore.WatchSchema != datastore.WatchSchema {
+		return
+	}
+
+	record := ch.recordForRevision(rev)
+	delete(record.definitionsChanged, "caveat-"+caveatName)
```

Comment:
use smaller prefix. Once this gets exposed via the API, there could be many consumers of the Watch API, and it's more bytes allocated unnecessarily